### PR TITLE
Fix warning message for matching on redundant nulls

### LINKED ISSUE #21724: Inaccurate warning message for the last redundant wildcard case in pattern match
## Compiler version

3.6.0-RC1-bin-SNAPSHOT-nonbootstrapped-git-6ceaab5

## Minimized code

```Scala
trait A
trait B

def test(x: A | B) = x match
  case _: A => println("A")
  case null => println("null")
  case _: B => println("B")
  case _ => println("default") // unreachable
```

## Output

The last case is reported as unreachable except for null, and suggested to be rewritten to `case null =>`.

```scala
-- [E121] Pattern Match Warning: Stest.scala:1240:7 ----------------------------
1240 |  case _ => println("default")
     |       ^
     |Unreachable case except for null (if this is intentional, consider writing case null => instead).
1 warning found
```

If we replace the wildcard to literal `null` in the last case, it will be reported as unreachable directly.

```scala
-- [E030] Match case Unreachable Warning: Stest.scala:1240:7 -------------------
1240 |  case null => println("default")
     |       ^^^^
     |       Unreachable case
1 warning found
```

## Expectation

The last redundant wildcard case should be reported as unreachable.

## Issue

In `checkReachability`:

```scala
if pat != EmptyTree // rethrow case of catch uses EmptyTree
    && !pat.symbol.isAllOf(SyntheticCase, butNot=Method) // ExpandSAMs default cases use SyntheticCase
    && isSubspace(covered, prev)
then {
  val nullOnly = isNullable && i == len - 1 && isWildcardArg(pat)
  val msg = if nullOnly then MatchCaseOnlyNullWarning() else MatchCaseUnreachable()
  report.warning(msg, pat.srcPos)
}
```

### PR DESCRIPTION
**Pull request by HarrisL2**
Fixes #21724

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 1]
```diff
@@ -937,12 +943,16 @@ object SpaceEngine {
...
                 && isSubspace(covered, prev)
             then
-              val nullOnly = isNullable && rest.isEmpty && isWildcardArg(pat)
+              val nullOnly = isNullable && isWildcardArg(pat) && !mayCoverNull(prev)
               val msg = if nullOnly then MatchCaseOnlyNullWarning() else MatchCaseUnreachable()
               report.warning(msg, pat.srcPos)
 
             // in redundancy check, take guard as false in order to soundly approximate
-            val newPrev = if guard.isEmpty then covered :: prevs else prevs
+            val newPrev = if (guard.isEmpty)
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:951]: I'm wondering whether we can modify the `project` to avoid iterating over space. For wildcard pattern with nullable types, we project to a nullable space.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 2]
```diff
@@ -912,6 +912,12 @@ object SpaceEngine {
     && !sel.tpe.widen.isRef(defn.QuotedExprClass)
     && !sel.tpe.widen.isRef(defn.QuotedTypeClass)
 
+  def mayCoverNull(tp: Space)(using Context): Boolean = tp match
+    case Empty => false
+    case Prod(_, _, _) => false
+    case Typ(tp, decomposed) => tp == ConstantType(Constant(null))
```
[the remark this answers is not in the document]
Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:918]: I get this error when writing `case _: Null` in a pattern match, so I did not add a NullType check. How should I test for it?
```
-- Error: tests/warn/redundant-null.scala:40:7 ---------------------------------
40 |  case _: Null => 
   |       ^^^^^^^
   |       class Null cannot be used in runtime type tests
```

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 1 above)
Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:951]: I will try this approach, thanks.

Comment (dwijnand) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:951]: > I'm wondering whether we can modify the `project` to avoid iterating over space. For wildcard pattern with nullable types, we project to a nullable space.

Can you reword that, please?  I didn't understand either sentence.. 👉🏼 👈🏼

Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:951]: I interpreted it as when the pattern is nullable, we project to a space with null included, like `targetSpace` on line 924.

**Comment by noti0na1** [comment]
Maybe I need to clarify my suggestion. I mean we always project a wildcard pattern to nullable space if the type is nullable.

For example: `case a =>` and `case _ =>` both will match the `null` value in runtime. This change should also fix some bug in the existing test suit, and make the redundant check more cleaner.

Like in `tests/warn/i20122.scala`:

```scala
sealed trait T_B[C, D]

case class CC_A()
case class CC_B[A, C](a: A) extends T_B[C, CC_A]
case class CC_C[C, D](a: T_B[C, D]) extends T_B[Int, CC_A]
case class CC_E(a: CC_C[Char, Byte])

val v_a: T_B[Int, CC_A] = CC_B(CC_E(CC_C(null)))
v_a match
  case CC_B(CC_E(CC_C(_))) => println(0) // was warn: unreachable
  case _                   => println(1)
```

If we run the code, we can see the first case is executed, so it is not unreachable. 


Then for the `OnlyNull` warning, I propose the following scheme. Issue OnlyNull warning only if:
*The target space is nullable;
* OnlyNull warning has not been issued before;
* The pattern is a wildcard pattern;
* The pattern is not covered by the previous cases, but covered by the previous cases with null.

I tried this approach, I think the code should be more cleaner: https://github.com/dotty-staging/dotty/blob/381c197ebbc1feb2f801c9df31d06e6d4d97fa6a/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala

Although this may contradict to what I said in `tests/explicit-nulls/warn/i21577.scala`...
```scala
def f3(s: String | Null) = s match
  case s2: String =>
  case _ => // warn NullOnly or not?
```

**Comment by noti0na1** [comment]
This list of tests should be corrected:

```
tests/warn/i20122.scala failed
tests/warn/i20121.scala failed
tests/warn/i20123.scala failed
```

Related to #21000

File: tests/warn/redundant-null.scala [hunk 3]
```diff
@@ -0,0 +1,41 @@
+class A
+class B
+class C
+
+val n = null
+
+def f(s: A) = s match
+  case _: n.type =>
+  case _: A =>
+  case _: n.type => // warn
```
Comment (noti0na1) [inline tests/warn/redundant-null.scala:10]: Can you also add the warning name in the test code? like `// warn: unreachable` and `// warn: null only`, etc.

Comment (HarrisL2) [reply tests/warn/redundant-null.scala]: Done

**Comment by noti0na1** [comment]
Maybe you need a rebase?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 4]
```diff
@@ -525,14 +533,25 @@ object SpaceEngine {
...
+        scrutineeTp match
+        case AppliedType(tycon, targs)
+            if unappSym.is(Synthetic)
+            && (pt.resultType.asInstanceOf[MethodType].paramInfos.head.typeConstructor eq tycon) =>
+          // Special case synthetic unapply/unapplySeq's
+          // Provided the shapes of the types match:
+          // the scrutinee type being unapplied and
+          // the unapply parameter type
+          pt.instantiate(targs).asInstanceOf[MethodType]
+        case _ =>
```
Comment (noti0na1) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala:545]: Is this some bug of github? This change has already been merged in main, why can I still see it here?

**Review by noti0na1 — APPROVED** [review]
Except the weird diff, otherwise LGTM. Can you try to `git rebase scala3/main` to see if you can get rid of the issue?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 4 above)
Comment (HarrisL2) [reply compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Rebased.

**Review by dwijnand — APPROVED** [review]
Looks great!  Thanks.
